packages/dd-trace/src/appsec/rasp/ssrf.js

```diff
@@ -25,16 +36,69 @@ function analyzeSsrf (ctx) {
 
   if (!req || !outgoingUrl) return
 
+  // Determine if we should collect the response body based on sampling rate and redirect URL
+  ctx.shouldCollectBody = downstream.shouldSampleBody(req, outgoingUrl)
+
+  const requestAddresses = downstream.extractRequestData(ctx)
+
   const ephemeral = {
-    [addresses.HTTP_OUTGOING_URL]: outgoingUrl
+    [addresses.HTTP_OUTGOING_URL]: outgoingUrl,
+    ...requestAddresses
   }
 
-  const raspRule = { type: RULE_TYPES.SSRF }
+  const raspRule = { type: RULE_TYPES.SSRF, variant: 'request' }
 
   const result = waf.run({ ephemeral }, req, raspRule)
 
-  const res = store?.res
-  handleResult(result, req, res, ctx.abortController, config, raspRule)
+  handleResult(result, req, store?.res, ctx.abortController, config, raspRule)
+
+  downstream.incrementDownstreamAnalysisCount(req)
+}
+
+/**
+ * Finalizes body collection for the response and triggers RASP analysis.
+ * @param {{
+ *   ctx: object,
+ *   res: import('http').IncomingMessage,
+ *   body: string|Buffer|null
+ * }} payload event payload from the channel.
+ */
+function handleResponseFinish ({ ctx, res, body }) {
+  if (!res) return
+
+  const store = storage('legacy').getStore()
+  const req = store?.req
+  if (!req) return
+
+  const evaluateBody = ctx.shouldCollectBody && downstream.handleRedirectResponse(req, res)
+
+  if (evaluateBody) {
+    // Skip body analysis for redirect responses
+    runResponseEvaluation(res, req, null)
+  } else {
+    runResponseEvaluation(res, req, body)
+  }
+}
+
+/**
+ * Evaluates the downstream response and records telemetry.
+ * @param {import('http').IncomingMessage} res outgoing response object.
+ * @param {import('http').IncomingMessage} req originating inbound request.
+ * @param {string|Buffer|null} responseBody collected downstream response body
+ */
+function runResponseEvaluation (res, req, responseBody) {
+  const responseAddresses = downstream.extractResponseData(res, responseBody)
+
+  if (!Object.keys(responseAddresses).length) return
+
+  const raspRule = { type: RULE_TYPES.SSRF, variant: 'response' }
+  const result = waf.run({ ephemeral: responseAddresses }, req, raspRule)
+
+  const ruleTriggered = !!result?.events?.length
+
+  if (ruleTriggered) {
+    updateRaspRuleMatchMetricTags(req, raspRule, false, false)
+  }
```

Comment:
```suggestion
  // TODO: this should be done in the waf functions directly instead of calling it everywhere
  const ruleTriggered = !!result?.events?.length

  if (ruleTriggered) {
    updateRaspRuleMatchMetricTags(req, raspRule, false, false)
  }
```